libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -346,6 +346,34 @@ string YulUtilFunctions::overflowCheckedUIntAddFunction(size_t _bits)
 	});
 }
 
+string YulUtilFunctions::overflowCheckedUIntMulFunction(size_t _bits)
+{
+	solAssert(0 < _bits && _bits <= 256 && _bits % 8 == 0, "");
+	string functionName = "checked_mul_uint_" + to_string(_bits);
+	return m_functionCollector->createFunction(functionName, [&]() {
+		return
+			// - The current overflow check *before* the multiplication could
+			//   be replaced by the following check *after* the multiplication:
+			//   if and(iszero(iszero(x)), iszero(eq(div(product, x), y))) { revert(0, 0) }
+			// - The case the x equals 0 could be treated separately and directly return zero.
+			Whiskers(R"(
+			function <functionName>(x, y) -> product {
+				if and(iszero(iszero(x)), lt(div(<mask>, x), y)) { revert(0, 0) }
```

Comment:
Why double iszero?

Reply:
Because if ``x`` is zero, we don't want to revert, so the condition for revert is ``x != 0`` and the other conjunct :-)? And I can't use ``and(x,...)``, since ``x`` doesn't necessarily have it's lowest bit set...

Reply:
I guess it could be ``gt(x,0)``, if that's better...

Reply:
Especially, since that'll be optimized to ``iszero(iszero(x))`` anyways.

Reply:
oh wow - I just actually was of the impression that it is logical and...